src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNet/AspNetCoreMvc2Integration.cs

```diff
@@ -0,0 +1,324 @@
+                Log.ErrorException($"Error retrieving {methodDef}", ex);
+                throw;
+            }
+
+            AspNetAmbientContext ambientContext = null;
+            var exceptionToGrab = context.GetProperty<Exception>("Exception");
+
+            if (shouldTrace)
+            {
+                var httpContextResult = context.GetProperty("HttpContext");
+
+                if (httpContextResult.HasValue)
+                {
+                    ambientContext = AspNetAmbientContext.RetrieveFromHttpContext(httpContextResult.Value);
+                }
+
+                if (ambientContext == null)
+                {
+                    Log.Error($"Could not access {nameof(AspNetAmbientContext)} for {methodDef}.");
+                }
+            }
+
+            try
+            {
+                // call the original method, catching and rethrowing any unhandled exceptions
+                instrumentedMethod(context);
+            }
+            catch (Exception ex) when (ambientContext?.SetExceptionOnRootSpan(exceptionToGrab.HasValue ? exceptionToGrab.Value : ex) ?? false)
+            {
+                // unreachable code
+                throw;
+            }
+        }
+
+        private static void SetAspNetCoreMvcSpecificData(
+            AspNetAmbientContext ambientContext,
+            object actionDescriptor,
+            object httpContext)
+        {
+            try
+            {
+                var request = httpContext.GetProperty("Request").GetValueOrDefault();
+
+                GetTagValues(
+                    actionDescriptor: actionDescriptor,
+                    request: request,
+                    httpMethod: out string httpMethod,
+                    resourceName: out string resourceName,
+                    controllerName: out string controllerName,
+                    actionName: out string actionName);
+
+                ambientContext.ResetWebServerRootTags(
+                    resourceName: resourceName,
+                    method: httpMethod);
+
+                var aspNetCoreMvcActionScope = ambientContext.Tracer.StartActive(OperationName);
+
+                aspNetCoreMvcActionScope.Span?.DecorateWebServerSpan(
+                    resourceName: resourceName,
+                    method: httpMethod,
+                    host: ambientContext.RootSpan?.GetHost(),
+                    httpUrl: ambientContext.RootSpan?.GetAbsoluteUrl());
+
+                aspNetCoreMvcActionScope.Span?.SetTag(Tags.AspNetController, controllerName);
+                aspNetCoreMvcActionScope.Span?.SetTag(Tags.AspNetAction, actionName);
+
+                ambientContext.TryPersistScope(IntegrationName, aspNetCoreMvcActionScope);
+
+                // set analytic sample rate if enabled
+                var analyticSampleRate = ambientContext.Tracer.Settings.GetIntegrationAnalyticsSampleRate(IntegrationName, enabledWithGlobalSetting: true);
```

Comment:
Are we doing this twice? Looks like we are already doing this in the `AspNetAmbientContext`.

Reply:
For different integrations technically.
"HttpContext" versus "AspNetCoreMvc2"
Should we only do this for top level?

Reply:
Oh also, I thought I was setting it on the aspNetCoreMvcSpan, not the root span, so either way, it's wrong.